modules/redis/redis.go

```diff
@@ -0,0 +1,267 @@
+	"github.com/netdata/go-orchestrator/module"
+)
+
+func init() {
+	creator := module.Creator{
+		Defaults: module.Defaults{
+			Disabled: false, // @TODO change it back
+		},
+		Create: func() module.Module { return New() },
+	}
+
+	module.Register("redis", creator)
+}
+
+const (
+	defaultHost = "localhost"
+	defaultPort = 6379
+)
+
+// New returns a pointer to a Redis instance with default values.
+func New() *Redis {
+	return &Redis{
+		charts: charts.Copy(),
+		Config: Config{
+			Host:     defaultHost,
+			Port:     defaultPort,
+			Password: "",
+			DbNum:    0,
+		},
+	}
+}
+
+// Config for the Redis module
+type Config struct {
+	Host     string `yaml:"host"`
+	Port     int    `yaml:"port"`
+	Password string `yaml:"password"`
+	DbNum    int    `yaml:"dbnum"`
+}
+
+// Redis module
+type Redis struct {
+	module.Base
+	charts      *Charts
+	client      *redis.Client
+	bgSaveTime  int64
+	Config      `yaml:",inline"`
+	UpdateEvery int `yaml:"update_every"`
+}
+
+// Cleanup closes the connection to the Redis server
+func (r *Redis) Cleanup() {
+	if err := r.client.Close(); err != nil {
+		r.Errorf("cleanup: error on closing the redis client [%+v]: %v", r.Config, err)
+	}
+}
+
+// Init checks the configuration parameters and pings the Redis server
+func (r *Redis) Init() bool {
+	if r.Host == "" {
+		r.Error("Host is not set")
+		return false
+	}
+
+	if r.Port == 0 {
+		r.Error("Port is not set")
+		return false
+	}
+
+	client := redis.NewClient(&redis.Options{
+		Addr:     r.Host + ":" + strconv.Itoa(r.Port),
+		Password: r.Password,
+		DB:       r.DbNum,
+	})
+
+	_, err := client.Ping().Result()
+	if err != nil {
+		r.Errorf("init: error on starting the redis client [%+v]: %v", r.Config, err)
+		return false
+	}
+
+	r.client = client
+	r.bgSaveTime = 0
+
+	r.Debugf("using config %+v", r.Config)
+
+	return true
+}
+
+// Check makes sure that at least one metric is being collected
+func (r *Redis) Check() bool {
+	return len(r.Collect()) > 0
+}
+
+// Charts returns the Redis charts
+func (r *Redis) Charts() *Charts {
+	return r.charts
+}
+
+// Collect returns a map of metrics
+func (r *Redis) Collect() map[string]int64 {
+	metrics := make(map[string]int64)
+
+	res, err := r.client.Do("INFO").Result()
+	if err != nil {
+		r.Errorf("Could not get Redis INFO: %v", err)
+		return nil
+	}
+
+	err = parseMetrics(res.(string), r.charts, metrics, &r.bgSaveTime, r.UpdateEvery)
+	if err != nil {
+		r.Errorf("Got error while parsing metrics: %v", err)
+		r.Debugf("INFO: %+v", res)
+	}
+
+	r.Debugf("Metrics: %+v", metrics)
+
+	return metrics
+}
+
+var autoParseDims = []string{
+	"total_commands_processed", "instantaneous_ops_per_sec",
+	"used_memory", "used_memory_lua",
+	"total_net_input_bytes", "total_net_output_bytes",
+	"evicted_keys",
+	"total_connections_received", "rejected_connections",
+	"connected_clients", "blocked_clients",
+	"connected_slaves",
+	"rdb_changes_since_last_save",
+	"uptime_in_seconds",
+}
+
+func parseMetrics(
```

Comment:
i think better to have the function as method, it doesnt look good to me how it is implemented right now:
 - a lot of input parameters
 - output of function actually is input (changes metrics and charts)
 - function name is parse, but it does more that it states - it parses, updates metric map, updates charts, etc.

What do you think?

Reply:
I agree on almost everything, let's break this down.

* output of function actually is input
   * I sometimes do that to avoid gc pressure but since the metrics map would escape to the heap anyway in the `Collect()` method we wouldn't avoid it anyway so I can definitely make `parseMetrics` initialize and return the map instead of having it as an input
* a lot of input parameters / function name is parse, but it does more that it states
   * I think both of these points you make could get down to the same thing, `parseMetrics` has too many responsibilities. I'd like to buy some time here and see how it looks like once it lives in the `collect.go` file and I add the Pika support. With Pika it should grow enough in complexity to give me an idea of how to better split this up.

About it being a method instead of a function I think I prefer it to be a function as it simplifies the tests. By having an unexported function wrapped in an exported method we could still expose a really nice API to whoever uses our package and at the same time we have a small unit (the private unexported function) that is super easy to test. More on this approach [here](https://stackoverflow.com/a/48206430/828366) if you can spare the time I'd like to know what you think about it.

Reply:
I think we need to use interface as `client` instead of `*redis.Client` and mock it during tests, otherwise you have to test `parseMetrics`. Testing parseMetrics != testing Collect (we are testing implementation details).

Reply:
> a really nice API to whoever uses our package

it is not a library, our modules are specific to netdata, their output has no sense outside netdata (unfortunately)

Reply:
oh, i see the same problem in the your approach 

> To summarize: test the unexported functions instead of testing the exported ones!

Maybe i am missing something, for me the problem is testing the details instead of logic, coupling logic with implementation.